src/Http/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs

```diff
@@ -0,0 +1,60 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.OpenApi.Models;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.OpenApi;
+
+/// <summary>
+/// Extension methods for annotating OpenAPI descriptions on an <see cref="Endpoint" />.
+/// </summary>
+public static class OpenApiRouteHandlerBuilderExtensions
+{
+    /// <summary>
+    /// Adds an OpenAPI annotation to <see cref="Endpoint.Metadata" /> associated
+    /// with the current endpoint.
+    /// </summary>
+    /// <param name="builder">The <see cref="RouteHandlerBuilder"/>.</param>
+    /// <param name="configureDescription">An <see cref="Action"/> that mutates an OpenAPI annotation.</param>
+    /// <returns>A <see cref="RouteHandlerBuilder"/> that can be used to further customize the endpoint.</returns>
+    public static RouteHandlerBuilder WithApiDescription(this RouteHandlerBuilder builder, Action<OpenApiPathItem>? configureDescription = null)
+    {
+        builder.Add(endpointBuilder =>
+        {
+            if (endpointBuilder is RouteEndpointBuilder routeEndpointBuilder)
+            {
+                var pattern = routeEndpointBuilder.RoutePattern;
+                var metadata = new EndpointMetadataCollection(routeEndpointBuilder.Metadata);
+                var methodInfo = metadata.OfType<MethodInfo>().SingleOrDefault();
+                var serviceProvider = routeEndpointBuilder.Metadata.OfType<IServiceProvider>().SingleOrDefault();
```

Comment:
This is kinda jank. The IServiceProvider shouldn't be metadata. I guess there's no other way to get the `IServiceProvider`?  Can we defer the need for it later?

Reply:
> I guess there's no other way to get the IServiceProvider?

Not that I can reason about.

> Can we defer the need for it later?

There's some alternatives. We mostly need DI to access the services we need to determine what parameters are what. We can punt on checking those parameters and make a parameter-less constructor for `OpenApiGenerator` that we instantiate directly here without relying on DI.

But I think we have to solve that problem eventually...